lib/autoBlock.js

```diff
@@ -141,41 +215,62 @@ function finish (error, results, config) {
 // Build out an async.auto block using our special notation
 function buildBlock (block, config) {
   var autoBlock = _.transform(block, (asyncBlock, step, key) => {
-    // Figure out which steps need to come before this one
-    // - anything in .after
-    // - the first path segement in each .with
-    // - the first path segement in each .when
-    var deps = []
-
-    var withDeps = _.map(step.with, (value) => _.first(_.split(value, '.')))
-
-    var whenDeps = _.map(step.when, (value, key) => {
-      var dep = _.isString(key) ? key : value
-      dep = _.first(_.split(dep, '.'))
-      if (_.first(dep) === '!') {
-        return dep.substring(1)
-      }
-
-      return dep
-    })
+    asyncBlock[key] = buildStep(step, config)
+  }, {})
+
+  return autoBlock
+}
+
+function buildStep (step, config) {
+  // Figure out which steps need to come before this one
+  // - anything in .after
+  // - the first path segement in each .with
+  // - the first path segement in each .when
+  var deps = []
 
-    deps = _.union(step.after, withDeps, whenDeps)
+  var withDeps = _.map(step.with, (value) => _.first(_.split(value, '.')))
 
-    // Build the async callback. This takes two forms: If no dependencies, no results paramater is included.
-    var asyncFunc
-    if (deps.length) {
-      asyncFunc = (results, cb) => runStep(results, cb, step, config)
-    } else {
-      asyncFunc = (cb) => runStep(null, cb, step, config)
+  var whenDeps = _.map(step.when, (value, key) => {
+    var dep = _.isString(key) ? key : value
```

Comment:
This is so `step.when` can be an object or an array? What can the values be if its an object?